### PR TITLE
Add flow control support on client and server.

### LINKED ISSUE #38: No throttling of message sending
Neither client nor server support flow control when sending messages (they do request
messages, so will interoperate with other servers and clients). This is because they
do not check `call.isReady`, or set the `onReady` handler.

Without this, the code which generates messages will not be throttled, and the code may
OOM in some situations.

### PR DESCRIPTION
**Pull request by fiadliel**
In order to support gRPC's flow control, the service has to handle
`call.isReady` and the `onReady` callback. When `isReady` is false,
the code should wait for `onReady` to be true. To avoid race conditions
where we may miss a wakeup, the `isReady` condition is checked twice;
once before, and once after assigning the wakeup `Deferred` value.

Fixes #38

### CODE REVIEW

**Comment by fiadliel** [comment]
@rossabaker @ahjohannessen 
I haven't done any testing yet, but opinions welcome. Also, do we want this in 0.4.0, or leave for afterwards? (given this is an existing issue).

**Comment by fiadliel** [comment]
Thinking about it, there are enough subtleties here that it might be best to release without this, and bake it for a while a later version.

**Comment by ahjohannessen** [comment]
This is a great addition, but perhaps, as you say, best to let it bake for a while and add tests.

**Comment by rossabaker** [comment]
My big itch was to have a release on fs2-1.0.  We've achieved that with the milestones, so I don't have strong opinions on whether this is makes it into 0.4.0 or a subsequent release.

**Comment by rossabaker** [comment]
This has fallen a bit behind.  Is this still something we want?

**Comment by ahjohannessen** [comment]
@rossabaker When @fiadliel says this

>Thinking about it, there are enough subtleties here that it might be best to release without this

it makes me a bit nervous with respect to stability and correctness. Perhaps tests and some canaries in the wild are needed.

**Comment by ahjohannessen** [comment]
@fiadliel I think this is important to support. What is needed to move this forward and what are the subtleties?

File: java-runtime/src/main/scala/server/Fs2ServerCall.scala [hunk 1]
```diff
@@ -16,11 +31,24 @@ private[server] class Fs2ServerCall[F[_], Request, Response](val call: ServerCal
   def sendMessage(message: Response)(implicit F: Sync[F]): F[Unit] =
     F.delay(call.sendMessage(message))
 
+  def sendMessageOrDelay(message: Response)(implicit F: Concurrent[F]): F[Unit] =
+    isReady.ifM(
+      sendMessage(message), {
+        Deferred[F, Unit].flatMap { wakeup =>
+          wakeOnReady.set(wakeup.some) *>
+            isReady.ifM(sendMessage(message), wakeup.get *> sendMessage(message))
```
Comment (flipp5b) [inline java-runtime/src/main/scala/server/Fs2ServerCall.scala:39]: Shouldn't we check `isReady` after `wakeup.get` completes to deal with missed wakeup?

**Review by flipp5b — COMMENTED** [review]
@fiadliel, first of all thank you for the great library and this PR!

I just want to throw in my two cents, maybe these comments would be useful.

**Comment by timbertson** [comment]
I'd love to see some progress on this work, any help needed? It seems like the code isn't huge, is the main remaining thing perhaps some extra tests?

**Comment by ahjohannessen** [comment]
@timbertson The code in this PR is outdated. There is also the question if #503 changes the status around flow control if it gets merged.

**Comment by timbertson** [comment]
Thanks @ahjohannessen . Based on https://github.com/typelevel/fs2-grpc/pull/503#discussion_r880950084 it sounds like backpressure will be unchanged in that branch.

I've opened #544 as an updated version of this PR, but targeted at v0.x.

**Comment by ahjohannessen** [comment]
Fixed by @timbertson in https://github.com/typelevel/fs2-grpc/pull/552 https://github.com/typelevel/fs2-grpc/pull/553

**Pull request closed without merging**